### PR TITLE
[feature](Nereids) Implement group by grouping sets, cube and rollup

### LINKED ISSUE #13615: support grouping sets in Nereids
The current new optimizer does not support related keywords such as grouping sets, as well as grouping and grouping_id functions. The purpose of this issue is to support related functions and syntax.
The related syntax and functions are used in the same way as the old optimizer, refer to:
https://doris.apache.org/zh-CN/community/design/grouping_sets_design

Subtasks for the issue:
- [x] Support group by grouping sets syntax #12583
- [ ] Support grouping and grouping_id functions #12583
- [ ] Support group by rollup and cube syntax
- [ ] Support group by rollup cube syntax

### PR DESCRIPTION
**Pull request by zhengshiJ**
# Proposed changes

Issue Number: close #13615 

The main work:

1. implement grouping sets/ cube/ rollup.
2. fix if function Infinite loop problem.
3. Support for isNull transitions to legacy optimizers.

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
4. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
5. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
6. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
7. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/SlotReference.java [hunk 1]
```diff
@@ -179,32 +167,28 @@ public SlotReference withChildren(List<Expression> children) {
...
-                    Optional.of(canPushDownPredicate), column);
+            return new SlotReference(exprId, name, dataType, nullable, qualifier, column);
         } else {
             boolean nullable = true;
-            boolean canPushDownPredicate = false;
-            return new SlotReference(exprId, name, dataType, nullable, qualifier,
-                    Optional.of(canPushDownPredicate), column);
+            return new SlotReference(exprId, name, dataType, nullable, qualifier, column);
         }
     }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/SlotReference.java:193]: you can delete this method and use `withNullable` method

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java [hunk 2]
```diff
@@ -586,4 +612,59 @@ public boolean hasGroupBy() {
             return false;
         }
     }
+
+    private List<NamedExpression> adjustNullableForProjects(
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:616]: leave behind some comment for adjestNullableXxx

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/If.java [hunk 3]
```diff
@@ -90,15 +84,6 @@ public class If extends ScalarFunction
...
-    private final Supplier<DataType> widerType = Suppliers.memoize(() -> {
-        List<AbstractDataType> argumentsTypes = getSignature().argumentsTypes;
-        Type assignmentCompatibleType = ScalarType.getAssignmentCompatibleType(
-                argumentsTypes.get(1).toCatalogDataType(),
-                argumentsTypes.get(2).toCatalogDataType(),
-                true);
-        return DataType.fromCatalogType(assignmentCompatibleType);
-    });
-
     /**
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/If.java:87]: wider type is nessesary, if there is some bug, you should fix it, but not delete this widerType

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 37.27 seconds
 load time: 491 seconds
 storage size: 17172817941 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221124023601_clickbench_pr_51770.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/SlotReference.java (hunk 1 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/SlotReference.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java (hunk 2 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindSlotReference.java:622]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/If.java (hunk 3 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/If.java:91]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/If.java [hunk 4]
```diff
@@ -90,27 +88,30 @@ public class If extends ScalarFunction
...
-        DataType widerType = this.widerType.get();
-        signature = signature.withArgumentTypes(arguments, (sigType, argType) -> widerType)
-                .withReturnType(widerType);
+        DataType widerType = getWiderType(signature.argumentsTypes);
+        List<AbstractDataType> newArgumentsTypes = new ImmutableList.Builder<AbstractDataType>()
+                .add(signature.argumentsTypes.get(0))
+                .add(widerType)
+                .add(widerType)
+                .build();
+        signature = signature.withArgumentTypes(signature.hasVarArgs, newArgumentsTypes);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/If.java:114]: missing set return type?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRepeat.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        return outputExpressions;
+    }
+
+    @Override
+    public String toString() {
+        return Utils.toSqlString("PhysicalRepeat",
+                "groupingSets", groupingSets,
+                "outputExpressions", outputExpressions
+        );
+    }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRepeat.java:99]:
```suggestion
    public String toString() {
        return Utils.toSqlString("PhysicalRepeat",
                "groupingSets", groupingSets,
                "outputExpressions", outputExpressions,
                "stats", statsDeriveResult
        );
    }
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/If.java (hunk 4 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/If.java:114]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRepeat.java (hunk 5 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRepeat.java:99]: done

**Review by 924060929 — APPROVED** [review]

**Pull request merged**